### PR TITLE
Add `DebuggerHidden` to managed JitHelpers

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Looking for thoughts on this before I move it out of draft.

/cc @dotnet/dotnet-diag @jkotas

### CODE REVIEW

**Comment by stephentoub** [comment]
What's the intended user experience in the debugger this is aiming to achieve?

**Comment by jkotas** [comment]
I do not think that this makes to apply wholesale to all throw helpers.

If this is necessary to make the recently introduce managed JIT throw helpers to behave more like the unmanaged JIT throw helpers, we should move the JIT throw helpers into a different type and only apply DebuggerHidden to methods in that type. There will be a bit of code duplication and that's fine.

**Comment by noahfalk** [comment]
Same thought as @stephentoub, its helpful to understand where these helpers are going to be used and what aspect of the debugging experience we are trying to change. My default principle is that what users see in the stack should match what they'd see reading the source code. So if the BCL C# code calls the helpers I'd prefer not to have those helpers hidden. If the JIT is inserting the helper where no such call exists in the source that would be a good place to hide it. I think @jkotas suggestion alludes to the same principle.

**Comment by AaronRobinsonMSFT** [comment]
This was suggested after trying to fix a diagnostic issue. We do seem slightly inconsistent when we apply this or not, so I just applied to all JitHelpers. The arguments presented by both @jkotas and @noahfalk make sense to me though. I figured it was appropriate to hide them in all cases. Seems that is not the case.

@tommcdon Did I misunderstand the suggestion?

**Comment by jkotas** [comment]
> I just applied to all JitHelpers

ThrowHelpers are not the same as JitHelpers.

**Comment by tommcdon** [comment]
> @tommcdon Did I misunderstand the suggestion?

https://github.com/dotnet/runtime/pull/105671 added an extra managed frame on to the stack for certain exceptions, regressing the developer experience.  For example with Just My Code off, we are now seeing a managed frame on the stack for IndexOutOfRangeException on array access:
```text
System.ThrowHelper.ThrowIndexOutOfRangeException (source line information unavailable)
ExceptionSpace.MyException._Generate_Exception0 (Exceptions.cs:59)
ExceptionSpace.MyException.Main (Exceptions.cs:29)
```
We are seeing the same behavior with DivideByZeroException on ARM.
I think it makes sense to hide these frames from the debugger as they are implementation detail to the runtime.  The debugger already hides native throw helpers in the runtime, and so I think it makes sense to hide throw helpers that were previously implemented in native code prior to https://github.com/dotnet/runtime/pull/105671.

**Comment by jkotas** [comment]
Right, I think the fix for this issue should be:
- Introduce a new type for the jit throw helpers. The throw helpers in this type should be only used as jit throw helpers, they should not be called directly from other CoreLib code as general purpose throw helpers
- Mark this new type and the methods in the type the same way as other managed jit helpers: https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/CastHelpers.cs
- Bonus points for deduplicating this new type with the native AOT JIT throw helpers (https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs). It does not need to be done in the same PR.

**Comment by AaronRobinsonMSFT** [comment]
> > I just applied to all JitHelpers
> 
> ThrowHelpers are not the same as JitHelpers.

Sure, that is fair. Several of the functions in this file are also JitHelpers. I naively applied it to all as if they were. I'll more narrowly apply `DebuggerHidden`.

> Introduce a new type for the jit throw helpers.

Will do.

**Comment by AaronRobinsonMSFT** [comment]
@jkotas I took your suggestion. I'm sure I missed something so this is likely to fail on native AOT or something, but can you take a look at the general direction?

**Comment by AaronRobinsonMSFT** [comment]
All green. @jkotas and @dotnet/dotnet-diag Please take another look.

**Review by tommcdon — APPROVED** [review]
LGTM, thanks Aaron!

File: src/coreclr/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs [hunk 1]
```diff
@@ -1,19 +1,23 @@
...
+using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
-namespace System
+namespace Internal.Runtime.CompilerHelpers
 {
-    internal static unsafe partial class ThrowHelper
+    [StackTraceHidden]
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs:11]: These attributes can be in the shared file.

File: src/libraries/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Internal.Runtime.CompilerHelpers
+{
+    /// <summary>
+    /// These methods are used to throw exceptions from generated code.
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs:13]: This ifdef is not necessary (left over from ProjectN). The comment and the ifdef can be deleted, both here and in the native AOT copy.

File: src/libraries/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    /// For NativeAOT, the type and methods used need to be public
+    ///     as they constitute a public contract.
+    /// For CoreCLR, the type and methods are used as JIT helpers.
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs:16]:
```suggestion
```
This ifdef is not necessary (left over from ProjectN). The comment and the ifdef can be deleted, both here and in the native AOT copy.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs:16]: (The type can be internal in native AOT, just like it is in CoreCLR. We have number of similar internal types in native AOT.)

**Review by jkotas — APPROVED** [review]
LGTM otherwise

**Review by noahfalk — COMMENTED** [review]
Oops, sorry I reviewed yesterday but apparently forgot to hit submit :(

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs [hunk 4]
```diff
@@ -11,7 +11,7 @@ namespace Internal.Runtime.CompilerHelpers
     /// These methods are used to throw exceptions from generated code. The type and methods
     /// need to be public as they constitute a public contract with the NativeAOT toolchain.
     /// </summary>
-    public static class ThrowHelpers
+    public static partial class ThrowHelpers
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs:14]: There is no difference (NativeAOT does not use the DebuggerHidden attribute currently). It is fine to make this consistent.

File: src/coreclr/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs [hunk 5]
```diff
@@ -1,19 +1,23 @@
...
+    [DebuggerStepThrough]
+    internal static unsafe partial class ThrowHelpers
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs:13]: `private` would not cause any problems with JIT usage. We have a bunch of existing managed JIT helpers that are `internal`, there is no consistency.

**Comment by AaronRobinsonMSFT** [comment]
Ugh. This is all sorts of messed up. I'll need to sit down and change more of this. There are API compat and protection levels issues when I change native AOT to `internal`.

For example:

> .../runtime/src/coreclr/nativeaot/System.Private.TypeLoader/src/Internal/TypeSystem/ThrowHelper.cs(4,55): error CS0122: 'ThrowHelpers' is inaccessible due to its protection level [.../runtime/src/coreclr/nativeaot/System.Private.TypeLoader/src/System.Private.TypeLoader.csproj]

**Comment by jkotas** [comment]
> runtime/src/coreclr/nativeaot/System.Private.TypeLoader/src/Internal/TypeSystem/ThrowHelper.cs(4,55): error CS0122: 'ThrowHelpers' is inaccessible due to its protection level

You can fix this by changing the NAOT TypeLoader to call RuntimeAugments, and move the throw helpers that are needed by the type loader there. 

I think Internal\Runtime\CompilerHelpers\ThrowHelpers.cs should be only used for ThrowHelpers that are called by the JITed code directly. It should not be used for cases where some random code needs to throw an exception.

**Comment by AaronRobinsonMSFT** [comment]
Updated.

**Review by jkotas — APPROVED** [review]
LGTM

**Pull request merged**